### PR TITLE
fix(datasource-mongoose): fix errors when creating or retrieving records from collections created with asModel on an object field

### PR DESCRIPTION
**Pull request by ghusse**
Linked to CU-860rq6mt1

### General

- [ ] Write an explicit title for the Pull Request, following [Conventional Commits specification](https://www.conventionalcommits.org)
- [ ] Test manually the implemented changes
- [ ] Validate the code quality (indentation, syntax, style, simplicity, readability)

### Security

- [ ] Consider the security impact of the changes made

### CODE REVIEW

File: packages/datasource-mongoose/src/collection.ts [hunk 1]
```diff
@@ -113,16 +113,31 @@ export default class MongooseCollection extends BaseCollection {
 
     // Only array fields can create subdocuments (the others should use update)
     const schema = MongooseSchema.fromModel(this.model).applyStack(this.stack);
-    if (!schema.isArray)
-      throw new ValidationError('Trying to create subrecords on a non-array field');
 
-    // Transform list of subrecords to a list of modifications that we'll apply to the root record.
+    if (schema.isArray) {
+      return this._createForArraySubfield(data, flatData, schema);
```
Comment (Scra3) [inline packages/datasource-mongoose/src/collection.ts:118]: You can remove the '_' before the private method. It's useless in typescript.

File: packages/datasource-mongoose/src/collection.ts [hunk 2]
```diff
@@ -113,16 +113,31 @@ export default class MongooseCollection extends BaseCollection {
...
+  }
+
+  private computeSubFieldName() {
     const lastStackStep = this.stack[this.stack.length - 1];
     const fieldName =
       this.stack.length > 2
         ? lastStackStep.prefix.substring(this.stack[this.stack.length - 2].prefix.length + 1)
         : lastStackStep.prefix;
 
+    return fieldName;
```
Comment (Scra3) [inline packages/datasource-mongoose/src/collection.ts:131]: Local variable 'fieldName' is redundant . You can return line 126.

File: packages/datasource-mongoose/src/collection.ts (hunk 1 above)
Comment (ghusse) [reply packages/datasource-mongoose/src/collection.ts:118]: Yes but I followed the convention in this file...

File: packages/datasource-mongoose/src/collection.ts [hunk 3]
```diff
@@ -146,17 +161,49 @@ export default class MongooseCollection extends BaseCollection {
...
   }
 
+  private async _createForObjectSubfield(data: RecordData[], flatData: RecordData[]) {
+    if (data.length > 1) throw new ValidationError('Trying to create multiple subrecords at once');
+
+    const fieldName = this.computeSubFieldName();
+
+    const entry = data[0];
+    const flatEntry = flatData[0];
+    const { parentId, ...rest } = entry;
```
Comment (Scra3) [inline packages/datasource-mongoose/src/collection.ts:184]:
```suggestion
    const { parentId, ...rest } = data[0];
```

File: packages/datasource-mongoose/src/collection.ts [hunk 4]
```diff
@@ -146,17 +161,49 @@ export default class MongooseCollection extends BaseCollection {
...
+        $set: {
+          [path]: rest,
+        },
+      },
+    );
+
+    return [
+      {
+        _id: `${rootId}.${path}`,
+        ...flatEntry,
```
Comment (Scra3) [inline packages/datasource-mongoose/src/collection.ts:202]:
```suggestion
        ...flatData[0],
```

File: packages/datasource-mongoose/src/collection.ts (hunk 2 above)
Comment (ghusse) [reply packages/datasource-mongoose/src/collection.ts:131]: Done

File: packages/datasource-mongoose/src/collection.ts (hunk 3 above)
Comment (ghusse) [reply packages/datasource-mongoose/src/collection.ts:182]: Done

File: packages/datasource-mongoose/src/collection.ts (hunk 1 above)
Comment (Scra3) [reply packages/datasource-mongoose/src/collection.ts:118]: Maybe you can remove all the '_' in the same PR.

File: packages/datasource-mongoose/src/collection.ts [hunk 5]
```diff
@@ -146,17 +161,49 @@ export default class MongooseCollection extends BaseCollection {
...
+        : this.model.updateOne({ _id: rootId }, { $set: { [path]: records[0] } }),
     );
 
     await Promise.all(promises);
 
     return results;
   }
 
+  private async _createForObjectSubfield(data: RecordData[], flatData: RecordData[]) {
+    if (data.length > 1) throw new ValidationError('Trying to create multiple subrecords at once');
```
[the remark this answers is not in the document]
Comment (ghusse) [reply packages/datasource-mongoose/src/collection.ts:176]: Done

File: packages/datasource-mongoose/src/collection.ts (hunk 1 above)
Comment (ghusse) [reply packages/datasource-mongoose/src/collection.ts:118]: Not in the same PR. I'll do it in a different one

**Review by Scra3 — APPROVED** [review]

**Pull request merged**